### ISSUE TITLE
Performance issue when drawing a rotated gradient into a clipped button

### ISSUE THREAD

**Issue by ened** [flutter/flutter#36935]
There seems to be a a severe spike in memory usage when drawing a slightly rotated gradient onto a `ClipPath` which contains a few child widgets. 

The rotation of the gradient is based on canvas clipping plus rotation and then explicitly overdrawing to create the desired effect in the target area.

The sample project below will easily rise to 500+ MB of RAM by scrolling the list up and down on a iPhone 6+, causing the app to run out of memory eventually.
If you scroll it fast enough you get the phone to reboot due to a memory fault.

## Steps to Reproduce

The sample project is available here: https://github.com/ened/flt_perf_03_rotated_gradient

![IMG_49ED6E24400D-1](https://user-images.githubusercontent.com/269860/61865222-c13ae500-aeca-11e9-9cfe-040b629374af.jpeg)

1. Clone sample project, adjust team
2. Run on iPhone 6+ using Xcode 
3. Scroll list up and down, watch the peak memory usage on the Debug Navigator
4. Press the "Enable Rotation" button in the App
5. Scroll list up and down again while watching the memory usage again - it will spike to over 500M

<img width="1183" alt="Screen Shot 2019-07-25 at 11 01 36" src="https://user-images.githubusercontent.com/269860/61865808-d19f8f80-aecb-11e9-9786-401c4e17c2ee.png">

## Logs

```
- no logs are generated -
```

```
$ flutter doctor -v
[✓] Flutter (Channel dev, v1.8.2, on Mac OS X 10.14.6 18G84, locale en-US)
    • Flutter version 1.8.2 at /Users/ened/dev/flutter
    • Framework revision 0a39d8d92e (2 weeks ago), 2019-07-10 19:17:14 -0700
    • Engine revision 75387dbc14
    • Dart version 2.5.0 (build 2.5.0-dev.0.0 b5aeaa6796)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /usr/local/opt/android-sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 28.0.3
    • ANDROID_HOME = /usr/local/opt/android-sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 10.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.3, Build version 10G8
    • CocoaPods version 1.7.4

[✓] iOS tools - develop for iOS devices
    • ios-deploy 1.9.4

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 37.1.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] VS Code (version 1.36.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.2.0

[✓] Connected device (2 available)
    • Sebastian Roth’s iPhone • 8abb0e857d6f964b7cb479db5f40cb70da7954b9 • ios • iOS 12.4
    • iPhone Xʀ               • 658FA040-9CEB-4A25-AC20-755A1AA9B5C2     • ios • com.apple.CoreSimulator.SimRuntime.iOS-12-4 (simulator)

• No issues found!
```

Would like to note that this bug is causing severe memory issues on older generation iPhones only. We had the rotated gradients in the App since version 1 and did not receive severe app crash reports on modern iPhones or Androids.

/cc @dnfield

**Comment by jason-simmons** [comment]
I just tried this on a Nexus 5x and a Moto G4.  The Nexus 5x showed a huge increase in graphics memory usage after clicking "enable rotation" and scrolling (eventually capping out at around 520MB).  Interestingly, the Moto G4 did not show a change in memory metrics.

**Comment by dnfield** [comment]
520 MB is the size of our Skia resource cache.

**Comment by ened** [comment]
#42484 seems to address the user task. Should we keep this issue open once it has been merged?

**Comment by dnfield** [comment]
That should allow you to rotate just the gradient, which should perform better.  You created this bug so you should be in a good position to determine if it is fixed :)

**Comment by dnfield** [comment]
For posterity, it would probably be nice to include a sample of what changes you made to resolve the performance issue though.

**Comment by ened** [comment]
@dnfield thank you for your work on this and especially #42484. 
I have updated my sample project to use it: https://github.com/ened/flt_perf_03_rotated_gradient/commit/884dfa6537b3094b2e680ea1c1c9b1adf51b8211.
There is a remaining problem. If I set the rotation to 90, then I'd expect the full gradient to be shown over the height of the button. However it seems like only a few steps (until the midway) are shown?
